snapcraft/remote/git.py

```diff
@@ -181,23 +184,69 @@ def push_url(
             "Pushing %r to remote %r with refspec %r.", ref, stripped_url, refspec
         )
 
-        # Create a list of tags to push
+        # temporarily call git directly due to libgit2 bug that unable to push
+        # large repos using https. See https://github.com/libgit2/libgit2/issues/6385
+        # and https://github.com/snapcore/snapcraft/issues/4478
+        cmd: list[str] = ["git", "push", remote_url, refspec, "--progress"]
         if push_tags:
-            tag_refs = [
-                t.name
-                for t in self._repo.references.iterator(pygit2.GIT_REFERENCES_TAGS)
-            ]
-        else:
-            tag_refs = []
+            cmd.append("--tags")
+
+        git_proc: Optional[subprocess.Popen] = None
         try:
-            self._repo.remotes.create_anonymous(remote_url).push([refspec] + tag_refs)
-        except pygit2.GitError as error:
+            with subprocess.Popen(
+                cmd,
+                cwd=str(self.path),
+                bufsize=1,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                universal_newlines=True,
+            ) as git_proc:
+                # do not block on reading from the pipes
+                # (has no effect on Windows until Python 3.12, so the readline() method is
+                # blocking on Windows but git will still proceed)
+                if git_proc.stdout:
+                    os.set_blocking(git_proc.stdout.fileno(), False)
+                if git_proc.stderr:
+                    os.set_blocking(git_proc.stderr.fileno(), False)
+
+                while git_proc.poll() is None:
+                    if git_proc.stdout:
+                        git_stdout: str = git_proc.stdout.readline()
+                        if git_stdout:
+                            logger.info(git_stdout.strip())
+                    if git_proc.stderr:
+                        git_stderr: str = git_proc.stderr.readline()
+                        if git_stderr:
+                            logger.error(git_stderr.strip())
```

Comment:
If we receive many lines to stdout or stderr at once, we probably want to output them all together.
```suggestion
                        while git_stderr: str = git_proc.stderr.readline():
                            logger.error(git_stderr.strip())
```

Reply:
That's a good idea